packages/@headlessui-react/src/utils/bugs.ts

```diff
@@ -0,0 +1,30 @@
+// See: https://github.com/facebook/react/issues/7711
+// See: https://github.com/facebook/react/pull/20612
+// See: https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#concept-fe-disabled (2.)
+export function isDisabledReactIssue7711(element: Element): boolean {
+  let parent = element.parentElement
+  let legend = null
+
+  while (parent && !(parent instanceof HTMLFieldSetElement)) {
+    if (parent instanceof HTMLLegendElement) legend = parent
+    parent = parent.parentElement
+  }
+
+  let isParentDisabled = parent?.getAttribute('disabled') === '' ?? false
+  if (isParentDisabled && isFirstLegend(legend)) return false
```

Comment:
Minor nitpick, but shouldn’t you return false if first legend no matter if in a disabled fieldset or not?

Reply:
It should result in the same behaviour:

```js
if (isParentDisabled && isFirstLegend(legend)) return false
return isParentDisabled
```

Because, when the `isParentDisabled` is false, then the result will be `false` anyway. So at this point it doesn't really matter that we are in a `legend` or not. The edge case is that when the parent **is** disabled **but** we are in the first legend, then we should ignore that.

Here is a truth table:

| isParentDisabled | isFirstLegend | Return Value | Info                                                                                                   |
| ---------------- | ------------- | ------------ | ------------------------------------------------------------------------------------------------------ |
| `true`           | `true`        | `false`      | Disabled, but first legend, so we can 'ignore' the disabled state.                                     |
| `true`           | `false`       | `true`       | Disabled, but not the first legend. `else` case is executed which is the result of `isParentDisabled`. |
| `false`          | `true`        | `false`      | Not disabled. `else` case is executed which is the result of `isParentDisabled`.                       |
| `false`          | `false`       | `false`      | Not disabled. `else` case is executed which is the result of `isParentDisabled`.                       |

I think that this is the correct behaviour 🤔

Reply:
Sorry for being unclear. I was on my phone with limited time (and UX...).

I wasn't implying it was incorrect (which is why I considered it a nitpick), merely that the code could possibly be _slightly_ simplified/refactored to:

```ts
if (isFirstLegend(legend)) return false
return isParentDisabled
```

Rationale: if it's in the first legend it should always return `false`, no matter where it is in the DOM or if it's enclosing `fieldset` is disabled.

Running it by your truth table yields the same result, as far as I can tell.

Not saying you should change it necessarily, you might find the current implementation more clear. Just throwing it out there how I see it.

Reply:
Ah yep, I see what you mean.

I tried to follow the wording of the spec (but it is not _that_ readable in the spec (2.))

> A form control is disabled if any of the following conditions are met:
> 
> 1. The element is a button, input, select, textarea, or form-associated custom element, and the disabled attribute is specified on this element (regardless of its value).
> 2. The element is a descendant of a fieldset element whose disabled attribute is specified, and is not a descendant of that fieldset element's first legend element child, if any.

But in the end, I don't think it matters much. Ideally we remove the code once React itself is fixed.